testsuites/store/src/main/resources/testcases/rdf-mt-inferencing/type/test003-in.nt

```diff
@@ -5,4 +5,6 @@
 # Relevant inferencing rules: rdfs10
 # Author: Arjohn Kampman (arjohn.kampman@aidministrator.nl)
 <http://example.org/prop> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2000/01/rdf-schema#ContainerMembershipProperty>.
-_:prop <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2000/01/rdf-schema#ContainerMembershipProperty>.
+
+# this entails that _:prop must be allowed as a predicate, but that is not legal in RDF
+#_:prop <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2000/01/rdf-schema#ContainerMembershipProperty>.
```

Comment:
You say you need to store "all the properties" which I assume means all resources that are (or can be) used as predicates, correct? If so, simply skip anything that isn't an IRI. Despite the fact that `_:prop` is defined as a property, it can never legally be _used_ as a property, so from a completeness point of view there's no need to keep track of it in the reasoner.

The point is that you need to distinguish between use and mention. It's perfectly legal to mention that `_:prop a rdf:Property`. So that test input is legal. What _isn't_ legal is for the reasoner to derive a _use_ of that fact where `_:prop` is actually used as a predicate.

And yes, that example you quote is illegal. It's part of the "generalized RDF"-section in the RDF Semantics spec, which is a theoretical discussion (non-normative) on "relaxing" RDF to allow blank nodes as predicates. That section shows why generalized RDF is required for rule-based deductive closure to be complete - the flipside of that is that by definition, any rule-based reasoner that sticks to the (non-generalized) RDF semantics is _by definition_ incomplete. In your example, `ex:d rdf:type ex:c` is a valid inference under RDF Semantics, but a rule-based reasoner cannot infer it because the intermediate result `ex:d _:b ex:c` is not legal RDF.